zxlive/settings_dialog.py

```diff
@@ -187,49 +184,75 @@ def __init__(self, main_window: MainWindow) -> None:
-            widget.setValue(int(val))  # type: ignore
-        elif ty == 'float':
-            widget = QDoubleSpinBox()
-            widget.setValue(float(val))  # type: ignore
-        elif ty == 'folder':
-            widget = QWidget()
-            hlayout = QHBoxLayout()
-            widget.setLayout(hlayout)
-            widget_line = QLineEdit()
-            widget_line.setText(str(val))
-            def browse() -> None:
-                directory = QFileDialog.getExistingDirectory(self,"Pick folder",options=QFileDialog.Option.ShowDirsOnly)
-                if directory:
-                    widget_line.setText(directory)
-                    setattr(widget, "text_value", directory)
-            hlayout.addWidget(widget_line)
-            button = QPushButton("Browse")
-            button.clicked.connect(browse)
-            hlayout.addWidget(button)
-        elif ty == 'combo':
-            widget = QComboBox()
-            assert data is not None
-            widget.addItems(list(data.values()))
-            widget.setCurrentText(data[val])
-            setattr(widget, "data", data)
-
-
-        form.addRow(label, widget)
-        self.value_dict[name] = widget
+    def make_str_form_input(self, data: SettingsData) -> QLineEdit:
+        widget = QLineEdit()
+        widget.setText(self.get_settings_from_data(data, str))
+        return widget
+
+    def make_int_form_input(self, data: SettingsData) -> QSpinBox:
+        widget = QSpinBox()
+        widget.setValue(self.get_settings_from_data(data, int))
+        return widget
+
+    def make_float_form_input(self, data: SettingsData) -> QDoubleSpinBox:
+        widget = QDoubleSpinBox()
+        widget.setValue(self.get_settings_from_data(data, float))
+        return widget
+
+    def make_folder_form_input(self, data: SettingsData) -> QWidget:
+        name = data["id"]
+        assert isinstance(default := defaults[name], str)
+        value = self.get_settings_value(name, str, default)
+        widget = QWidget()
+        hlayout = QHBoxLayout()
+        widget.setLayout(hlayout)
+        widget_line = QLineEdit()
+        widget_line.setText(value)
+
+        def browse() -> None:
+            directory = QFileDialog.getExistingDirectory(
+                self,"Pick folder", options=QFileDialog.Option.ShowDirsOnly
+            )
+            if directory:
+                widget_line.setText(directory)
+                setattr(widget, "text_value", directory)
+
+        hlayout.addWidget(widget_line)
+        button = QPushButton("Browse")
+        button.clicked.connect(browse)
+        hlayout.addWidget(button)
+        return widget
+
+    def make_combo_form_input(self, data: SettingsData) -> QComboBox:
+        name, _data = data["id"], data["data"]
+        value = self.settings.value(name, defaults[name])
+        widget = QComboBox()
+        widget.addItems(list(_data.values()))
+        widget.setCurrentText(_data[value])
+        setattr(widget, "data", _data)
```

Comment:
It might be better to subclass QComboBox instead of dynamically adding the `data` attribute.